pkg/scheduler/framework/plugins/imagelocality/image_locality.go

```diff
@@ -41,17 +41,67 @@ type ImageLocality struct {
 }
 
 var _ framework.ScorePlugin = &ImageLocality{}
+var _ framework.PreScorePlugin = &ImageLocality{}
 
-// Name is the name of the plugin used in the plugin registry and configurations.
-const Name = names.ImageLocality
+const (
+	// Name is the name of the plugin used in the plugin registry and configurations.
+	Name = names.ImageLocality
+
+	// preScoreStateKey is the key in CycleState to ImageLocality pre-computed data for Scoring.
+	preScoreStateKey = "PreScore" + Name
+)
+
+// preScoreState computed at PreScore and used at Score.
+type preScoreState struct {
+	containersPullAlways bool
+}
+
+// Clone implements the mandatory Clone interface. We don't really copy the data since
+// there is no need for that.
+func (s *preScoreState) Clone() framework.StateData {
+	return s
+}
 
 // Name returns name of the plugin. It is used in logs, etc.
 func (pl *ImageLocality) Name() string {
 	return Name
 }
 
+func (pl *ImageLocality) PreScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
+	pullAlways := true
+
+	// Filtering containers with ImagePullPolicy different from Always
+	// Ones with Always will score 0 point either way
+	for _, container := range pod.Spec.Containers {
+		if container.ImagePullPolicy != v1.PullAlways {
+			pullAlways = false
+			break
+		}
+	}
+
+	state := &preScoreState{
+		containersPullAlways: pullAlways,
+	}
+
+	cycleState.Write(preScoreStateKey, state)
+	return nil
+}
+
 // Score invoked at the score extension point.
 func (pl *ImageLocality) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	s, err := getPreScoreState(state)
+	if err != nil {
+		pl.PreScore(ctx, state, pod, []*v1.Node{})
+		s, err = getPreScoreState(state)
```

Comment:
Can we get the necessary information without writing to the state?  We can make finding the `imagePullPolicy=Always` container a method and recall here.

Reply:
When PreScore is not enabled, only the first node needs to call PreScore to obtain cycleState, and other nodes can directly read cycleState. If it is changed to a function, each node needs to be executed once